## 1. Demultiplexing raw fastq.gz files

* Used process_radtags: https://catchenlab.life.illinois.edu/stacks/comp/process_radtags.php

--renz = enzymes  
-c = clean data, remove any read with an uncalled base  
-q = discard reads with low quality scores  
-r = rescue barcodes and RAD-Tag cut sites  
default barcode option --inline_null: barcode is inline with sequence, occurs only on single-end read  
reads with phred-scores < an average of 10 (within a sliding window 0.15x the read length) were discarded by default

```
cd stacks-2.53

process_radtags -p filepath/raw_data/ -o filepath/clean_data/ --renz_1 sphI --renz_2 mluCI -c -q -r
```

* Remove unwanted samples from clean_data. Popmap files for all 4 species can be found here: [popmap_files](https://github.com/mae47/Red_Sea_LGM/tree/main/data/popmap_files)

## 2. Assemble de-novo loci

* Used the ‘denovo_map.pl’ pipeline run up to the creation of the bam files (ustacks, cstacks, sstacks, tsv2bam): https://catchenlab.life.illinois.edu/stacks/comp/denovo_map.php

* Options for building stacks were as described in Salas et al. (2019) https://royalsocietypublishing.org/doi/10.1098/rsos.172413
as below:

ustacks -M = 3, maximum number of mismatches allowed between stacks within individuals
ustacks -m = 3, minimum depth of coverage to create a stack 
ustacks –N = 5 ('M' + 2), maximum number of mismatches allowed to align secondary reads to primary stacks
cstacks -n = 2, maximum number of mismatches allowed between stacks between individuals

```
cd stacks-2.53/bin

#ustacks
#Build de-novo loci in each sample for the single-end reads only

id=1
for sample in SRR7226206 SRR7226207 SRR7226214 ....
do
ustacks -f filepath/clean_data/$sample.fq.gz -i $id -o filepath/D_trimaculatus/output_24ind -M 3 -m 3 -N 5
let "id+=1"
done

#cstacks
#Build the catalog of loci available in the metapopulation from the samples contained in the population map

cstacks -P filepath/D_trimaculatus/output_24ind -M filepath/popmap_Dt_24ind -n 2

#sstacks
#Match all samples supplied in the popmap against the catalog ie. sets of stacks (putative loci) can be searched against the catalog

sstacks -P filepath/D_trimaculatus/output_24ind -M filepath/popmap_Dt_24ind

#tsv2bam
#transpose the data so it is stored by locus instead of by sample. can handle paired-end

tsv2bam -P filepath/D_trimaculatus/output_24ind -M filepath/popmap_Dt_24ind
```

Resulting bam files were filtered for a minimum coverage of eight using samtools v.1.9 (Danecek et al., 2021)

before continuing with the ‘gstacks’ and ‘populations’ components of the ‘denovo_map.pl’ pipeline
