---
title: "Untitled"
author: "Maddie"
date: "26/02/2021"
output: html_document
---

#Calculate the 2d sfs (in relation to both RS and Sey popns), function vcf2sfs. Need vcf & pop file (ID, popn)

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



```{r}
#complete to automate later row/col folding: rowpop/pop2/Red Sea samples 
numrow<-14
#colpop/pop1/Indian Ocean samples
numcol<-14

#location of species folder. Make sure there is an empty subfolder created called "sfs_target"
filename<-"/home/maddie/Documents/PhD/LGM_paper/new_plots/sfs/C_melanopterus"
```


Create mygt, a genotype table, from diploid vcf and popmap. NB need to check correct file locations
0 - Homo 0/0
1 - Hetero 1/0 0/1
2 - Homo 1/1

```{r}
vcf2gt<-function(f.vcf,f.popmap)
{
	oldw<-getOption("warn")
	options(warn=-1)
	
	# Read VCF file and popmap file.
	vcf.gt<-as.matrix(read.table(f.vcf,sep="\t",stringsAsFactors=F)[,-c(1:9)])
	popmap<-read.table(f.popmap,sep="\t",stringsAsFactors=F)[,2]
	
	nrow.vcf<-nrow(vcf.gt)
	ncol.vcf<-ncol(vcf.gt)
	
	# Parse genotypes.
	chrom1<-substring(vcf.gt,1,1)
	chrom2<-substring(vcf.gt,3,3)
	chrom<-matrix(as.integer(chrom1)+as.integer(chrom2),nrow.vcf,ncol.vcf)
	
	options(warn=oldw)
	
	list(popmap=popmap,genotype=chrom)
}


print("function vcf2gt read")
#NB check popmap pops are sey/rs rather than 1/2 (always using sey even if other areas of Indian Ocean were sampled)
mygt<-vcf2gt(paste0(filename,"/populations.snps.vcf"),paste0(filename,"/popmap_Cm_28ind")) 
print("mygt created")
```


1/2
SHARK: generate SFS not accounting for missing data at all (already filtered out)

```{r}
## Generate a SFS (table object) from the gt object, not accounting for missing data
# It will output an imputed SFS, not accounting for the missing values.
# pops: a character or integer vector; IDs of populations to be included in the SFS.
gt2sfs.raw<-function(gt,pops)
{
	popmap<-gt$popmap
	chrom<-gt$genotype
	
	nrow.vcf<-nrow(chrom)
	ncol.vcf<-ncol(chrom)
	
	n.pop<-length(pops)
	
	# Number of chromosomes.
	ns.chr<-sapply(pops,function(x){sum(popmap==x)})*2
	
	# SFS based on raw count.
	cnt<-matrix(0,nrow.vcf,n.pop)
	ext<-list()
	for(i in 1:n.pop)
	{
		index<-which(popmap==pops[i])
		cnt[,i]<-rowSums(chrom[,index],na.rm=T)
		ext<-c(ext,list(0:ns.chr[i]))
	}
	ext<-as.matrix(expand.grid(ext))
	cnt<-data.frame(rbind(cnt,ext))
	sfs.raw<-table(cnt)-1
	names(dimnames(sfs.raw))<-pops
	
	sfs.raw
}


print("function gt2sfs.raw read")
#compute 2d-sfs
popns<-(c("rs","sey"))
mysfs<-gt2sfs.raw(mygt, popns)
print("mysfs created")
print(mysfs)
write.table(mysfs, paste0(filename,"/sfs_target/mysfs_NA0.txt"))
```


2/2
TELEOSTS: Generate SFS (Pier's 05032021 edited impute function) filling in ONLY the missing data

```{r}
###Need to load gt2sfs function first (see below, SFS method 3/3) for use in Pier's impute function 
gt2sfs.raw<-function(gt,pops)
{
	popmap<-gt$popmap
	chrom<-gt$genotype
	
	nrow.vcf<-nrow(chrom)
	ncol.vcf<-ncol(chrom)
	
	n.pop<-length(pops)
	
	# Number of chromosomes.
	ns.chr<-sapply(pops,function(x){sum(popmap==x)})*2
	
	# SFS based on raw count.
	cnt<-matrix(0,nrow.vcf,n.pop)
	ext<-list()
	for(i in 1:n.pop)
	{
		index<-which(popmap==pops[i])
		cnt[,i]<-rowSums(chrom[,index],na.rm=T)
		ext<-c(ext,list(0:ns.chr[i]))
	}
	ext<-as.matrix(expand.grid(ext))
	cnt<-data.frame(rbind(cnt,ext))
	sfs.raw<-table(cnt)-1
	names(dimnames(sfs.raw))<-pops
	
	sfs.raw
}


# It will output an imputed SFS accounting for the missing values through binomial distribution.
# Only missing genotypes will be imputed!!!
# pops: a character or integer vector; IDs of populations to be included in the SFS.
# sampleSizes: an integer vector; downsized number of individuals in each population.
gt2sfs.impute_pier<-function(gt,pops,sampleSizes)
{
  popmap<-gt$popmap
  chrom<-gt$genotype/2
  
  nrow.vcf<-nrow(chrom)
  ncol.vcf<-ncol(chrom)
  
  n.pop<-length(pops)
  
  # Number of chromosomes.
  ns.chr<-sampleSizes*2
  
  #create a genotype table where imputed genotypes will be added
  gt_imputed<-gt
  imputed_na_genotypes<-c()
 
  for(i in 1:n.pop)
  {
    index<-which(popmap==pops[i])
    p<-rowMeans(chrom[,index],na.rm=T)
    rows_with_na<-which(rowSums(is.na(chrom[,index]))!=0)
    
    #record the number of the lines with missing data (column_1), the associate probability to use for rbinom (column_2) and how many missing genotypes to generate (column_3)
    na_summary<-matrix(rep(0, 3*length(rows_with_na)), ncol=3, nrow=length(rows_with_na))
    na_summary[,1]<-rows_with_na
    na_summary[,2]<-p[rows_with_na]
    na_summary[,3]<-rowSums(is.na(chrom[,index]))[rowSums(is.na(chrom[,index]))!=0]
    
    #check if all genotypes are missing and therefore we cannot impute them
    if(any(is.na(na_summary[,2]))) {
      stop(paste("All genotypes for site number ", na_summary[which(is.na(na_summary[,2])),1], " in population ", pops[i]," are missing", sep=""))
    }
    
    #impute all missing data
    for (k in 1:nrow(na_summary)) {
      single_gen<-rbinom(na_summary[k,3],2,na_summary[k,2])
      imputed_na_genotypes<-append(imputed_na_genotypes, single_gen, after=length(imputed_na_genotypes))
    }
  }
  
  #substitute the missing data with the imputed genotypes
  gt_imputed$genotype[which(is.na(gt_imputed$genotype))]<-imputed_na_genotypes
  
  #use the function gt2sfs.raw to generate the sfs with imputed genotypes
  gt2sfs.raw(gt_imputed,pops)
}


print("gt2sfs.raw function read")
print("gt2sfs.impute_pier function read")
mysfs<-gt2sfs.impute_pier(mygt, c("rs","sey"), c(numrow,numcol))
print("mysfs created")
print(mysfs)
write.table(mysfs, paste0(filename,"/sfs_target/mysfs_impute.txt"))

```



Fold SFS:

```{r}
fold.sfs<-function(sfs)
{
		sfs[]<-sfs+rev(sfs)
	dims<-dim(sfs)
		cnt.pool<-rowSums(expand.grid(lapply(dims-1,function(x)0:x)))
		index<-cnt.pool>(sum(dims-1)/2)
			sfs[index]<-0
			index<-cnt.pool==(sum(dims-1)/2)
				sfs[index]<-sfs[index]/2
				sfs
}

mysfs_folded<-fold.sfs(mysfs)
print(mysfs_folded)
print(t(mysfs_folded))
##SHARK
write.table(mysfs_folded, paste0(filename,"/sfs_target/mysfs_NA0_folded.txt"))
##TELEOSTS
#write.table(mysfs_folded, paste0(filename,"/sfs_target/mysfs_impute_folded.txt"))


```


Plot SFS
NB With plotting script from vcf2sfs github page, I struggled with scaled raster legend. Now using ggplot2. 
PRIOR TO PLOTTING:
NB edit mysfs_impute_folded.txt/mysfs_NA0_folded.txt manually to contain an extra name in col. 1 of header line (eg "samples") to align rows and columns of table
NB for shark TGC data, also edit mysfs_NA0_folded.txt manually to remove monomorphic sites ie. the single number/value in column 2, row 2 ie. SNPs not present in either population (not an issue as not present in RADseq/stacks data)

```{r}
##Suggested ggplot script from Pier 
plots<-list()
library(ggplot2)
library(dplyr)
library(tidyr)
library(viridis)

##TELEOSTS
#t<-read.table(paste0(filename,"/mysfs_impute_folded.txt"), header=T,  sep=" ", check.names=F)
#t1<-gather(t, key="pop2", value="SNPs", -pop1) #gather columns into key-value pairs - used when columns are not variables (-pop1 variable name is excluded): 0-0...18-0,0-1...18-1,0-2...18-2,...
#g <- ggplot(data=t1, aes(pop1,y=sort(as.numeric(pop2)), fill=log10(SNPs))) + 
  ##NB pop2, numrow, rows in SFS table, should be labelled on x axis of sfs plots
  #geom_tile() + ggplot2::xlab("Red Sea") + ggplot2::ylab("Indian Ocean") +
  #scale_fill_viridis(labels=c(0,10,100,1000,10000),name="SNPs")
##Repeat this for Dt "B" [[2]] and Pm "C" [[3]]
#Da <- g + labs(tag="A")
#plots[[1]]<-Da
##ggsave(file="/home/maddie/Desktop/D_abudafur/mysfs_impute_folded.png", g)

#SHARK
t<-read.table(paste0(filename,"/sfs_target/mysfs_NA0_folded.txt"), header=T, sep=" ", check.names=F)
t1<-gather(t, key="pop2", value="SNPs", -pop1) #gather columns into key-value pairs - used when columns are not variables (-pop1 variable name is excluded): 0-0...18-0,0-1...18-1,0-2...18-2,...
g <- ggplot(data=t1, aes(pop1,y=sort(as.numeric(pop2)), fill=log10(SNPs))) + 
  ##NB pop2, numrow, rows in SFS table, should be labelled on x axis of sfs plots
  geom_tile() + ggplot2::xlab("Red Sea") + ggplot2::ylab("Indian Ocean") +
  scale_fill_viridis(labels=c(0,1,3,10,32,100,1000), name="SNPs")
  Cm <- g + labs(tag="D")
plots[[4]]<-Cm
##ggsave(file="/home/maddie/Desktop/C_melanopterus/mysfs_NA0_folded.png", g)
```


If saved all species ggplot objects, can plot all together here:
NB Must run whole block at once or will get "cannot shut down device 1 error"
```{r}
library(ggplot2)
library("gridExtra") #for grid.arrange and arrangeGrob below
library(ggpubr)

png("/home/maddie/Documents/PhD/LGM_paper/new_plots/sfs/sfs_multi.png",width=800,height=700)

g <- ggarrange(Da,Dt,Pm,Cm, nrow = 2, ncol=2,labels = c("A","B","C","D"))


print(g)
dev.off()
```


write to .dadi format:

```{r}
write.sfs.dadi<-function(sfs,f.output)
{
  sink(f.output)
  cat(dim(sfs))
  cat("\n")
  cat(aperm(sfs))
  sink()
}


print("function write.sfs.dadi read")
##TELEOSTS
#write.sfs.dadi(mysfs_folded, paste0(filename,"/sfs_target/mysfs_impute_folded_dadi.txt"))
##SHARK
write.sfs.dadi(mysfs_folded, paste0(filename,"/sfs_target/mysfs_NA0_folded_dadi.txt"))
print("dadi file created")

#NB remember to delete header now, and manually add between-pop pi at the end of the file at this point, if desired. Otherwise need to add an end of line character
```


Remove "cut" unwanted columns of MAF table from dadi file. NB need to adjust script depending on whether or not removing singleton sites:

```{r}
#call mysfs_folded_dadi.txt target file to variable
#mysfs_folded_dadi<-read.table(paste0(filename,"/sfs_target/mysfs_impute_folded_dadi.txt"))
mysfs_folded_dadi<-read.table(paste0(filename,"/sfs_target/mysfs_NA0_folded_dadi.txt"))
final_col<-ncol(mysfs_folded_dadi)
pi<-mysfs_folded_dadi[,final_col]


#developing string of columns to be kept
if (numcol>numrow || numcol==numrow) {    #the larger of the 2 pops is along the columns (sey) or both are same size
  numchr<-numcol*2    #no. chromosomes in the the larger pop
    
    if (numcol>numrow) {
    print("numcol > numrow") 
    } else {print("numcol = numrow")}
  
i<-1  #first iteration
start<-numrow+numcol+1  #maxCols for row0
#x<-c(seq((i+2),start))  #i+2 skips 0-0 and 0-1 categories (eliminating monomorphic and singleton categories)
#print(paste0("skipping monomorphic and singletons, row 1 printed up to column ", start, " of total columns ", numchr+1))
x<-c(seq((i+1),start))  #i+1 skips 0-0 category (eliminating monomorphic sites)
print(paste0("skipping monomorphic sites only, row 1 printed up to column ", start, " of total columns ", numchr+1))
print(x)

#second iteration for row2
i<-i+(numchr+1)
start<-start+numchr
#x2<-c(seq((i+1),start))  #i+1 skips 1-0 category (eliminating singleton category)
#print(paste0("skipping singletons, row 2 printed up to column ", numrs+numsey))
x2<-c(seq(i,start))  #not removing singleton category
print(paste0("retaining singleton category, row 2 printed up to column ", numrow+numcol))
x<-c(x,x2)
print(x)

#third iteration onwards (third row onwards)
for (val in 3:(numrow*2+1)) {  #all rows of smaller pop chromosomes inc. 0
    print(paste0("row ", val))
i<-i+(numchr+1)
start<-start+numchr
x2<-c(seq(i,start))
x<-c(x,x2)
}
print(paste0("is last row. Printed up to column ", numcol-numrow+1))
print("selected columns:")
print(x)
  #mysfs_folded_dadi[,x] 



}  else {            #the smaller of the 2 pops is along the columns (sey)
  numchr<-numcol*2    #no. chromosomes in the smaller pop
  print("numrow > numcol")
  
  i<-1  #first iteration
  start<-numchr+1
 #x<-c(seq((i+2),start))  #i+2 skips 0-0 and 0-1 categories (eliminating monomorphic and singleton categories)
  #print("skipping monomorphic and singletons, printed all columns of row 1")
  x<-c(seq((i+1),start))  #i+1 skips 0-0 category (eliminating monomorphic sites)
  print("skipping monomorphic sites only, printed all columns of row 1")
  print(x)

  #second iteration
i<-start+1
y<-((numrow-numcol)+1)   #number of rows (inc. 0) for which all smaller pop chromosome columns are kept
start<-((numchr+1)*y)  #sum of remaining rows for which all cols kept
#x2<-c(seq((i+1),start))  #i+1 skips 1-0 category (eliminating singleton category)
#print(paste0("skipping singletons, printed all columns from row 2, then all columns from row 3 to row ", y, ". Now starting to drop columns:"))
x2<-c(seq(i,start))  #not removing singleton category
print(paste0("retaining singleton category, printed all columns from row 2, then all columns from row 3 to row ", y, ". Now starting to drop columns:"))
x<-c(x,x2)
print(x)

#third iteration starting to cut columns off the end
for (val in 1:(numchr)) {   #all rows of smaller pop chromosomes inc. 0
print(paste0("row ", y+val))
    i<-start+val
start<-start+numchr
x2<-c(seq(i,start))
x<-c(x,x2)
}
print(paste0("is last single column row of total rows ", numrow*2+1))
print("selected columns:")
x
  #mysfs_folded_dadi[,x] 
}


mysfs_folded_dadi_2<-mysfs_folded_dadi[,x]
mysfs_folded_dadi_2<-cbind(mysfs_folded_dadi_2,pi)
dadi_cols<-ncol(mysfs_folded_dadi_2)
mysfs_folded_dadi_2[,(dadi_cols-1):dadi_cols]

##saving out cut dadi target
#write.table(mysfs_folded_dadi_2, paste0(filename,"/sfs_target/mysfs_impute_folded_dadi_cut_nosingletons.txt"), col.names=FALSE, row.names=FALSE, sep=" ", quote=FALSE)
#write.table(mysfs_folded_dadi_2, paste0(filename,"/sfs_target/mysfs_impute_folded_dadi_cut.txt"), col.names=FALSE, row.names=FALSE, sep=" ", quote=FALSE)
write.table(mysfs_folded_dadi_2, paste0(filename,"/sfs_target/mysfs_NA0_folded_dadi_cut.txt"), col.names=FALSE, row.names=FALSE, sep=" ", quote=FALSE)


```

 
